openstudiocore/src/model/EnergyManagementSystemProgram.cpp

```diff
@@ -0,0 +1,430 @@
+    if (result.empty()){
+    }
+    return result;
+  }
+
+  IddObjectType EnergyManagementSystemProgram_Impl::iddObjectType() const {
+    return EnergyManagementSystemProgram::iddObjectType();
+  }
+
+  boost::optional<std::string> EnergyManagementSystemProgram_Impl::body() const {
+    //return program body as string
+
+    // loop through extensible groups and add ProgramLine to body string.
+    std::string body;
+    boost::optional<std::string> comment;
+
+    auto groups = extensibleGroups();
+    for (auto group = groups.begin(); group != groups.end(); ++group) {
+      //get program line content
+      const auto & line = group->getString(OS_EnergyManagementSystem_ProgramExtensibleFields::ProgramLine, true);
+      OS_ASSERT(line);
+      body += line.get();
+      //get non-default comments if they exist
+      comment = group->fieldComment(OS_EnergyManagementSystem_ProgramExtensibleFields::ProgramLine, false);
+      if (comment.is_initialized()) {
+        //remove space after !
+        boost::erase_first(comment.get()," ");
+        //add space between end of program line and comment
+        if (!comment.get().empty()) {
+          body += " " + comment.get();
+        }
+      }
+      //add newline character
+      body += '\n';
+    }
+    return body;
+  }
+
+  bool EnergyManagementSystemProgram_Impl::setBody(const std::string& body) {
+    //set body of program to input string
+    bool result = false;
+    bool comment_result = false;
+    //if body string empty then return false
+    if (body.empty()) {
+      return false;
+    };
+
+    //clobber existing body
+    this->eraseBody();
+
+    // remove '\r' from the body string
+    std::string body_minus_r = body;
+    std::string::size_type pos = 0; // Must initialize
+    while ((pos = body_minus_r.find("\r", pos)) != std::string::npos) {
+      body_minus_r.erase(pos, 1);
+    }
+
+    //split the body string on newline characters and comments and insert program line for each string line
+    std::vector<std::string> body_minus_nl = splitString(body_minus_r, '\n');
+    std::string body_minus_comment;
+    std::vector<std::string> comments;
+    std::string comment;
+
+    //add program lines to body
+    std::vector<bool> ok(body_minus_nl.size(), false);
+    for (size_t i = 0; i < body_minus_nl.size(); i++) {
+      //split string on comment character !
+      comments = splitString(body_minus_nl.at(i), '!');
+      //get extensibleGroup object
+      WorkspaceExtensibleGroup group = getObject<ModelObject>().pushExtensibleGroup().cast<WorkspaceExtensibleGroup>();
+      //make sure program line exists and insert
+      if (comments.size() > 0) {
+        //remove whitespace at end of comments[0]
+        boost::trim_right(comments[0]);
+        //remove whitespace at beginning of comments[0]
+        boost::trim_left(comments[0]);
+        //remove ,
+        pos = 0;
+        while ((pos = comments[0].find(",", pos)) != std::string::npos) {
```

Comment:
comments[0] is the actual program line plus the , to indicate a new line.  I remove the , because the setString(...:ProgramLine) (line 150) adds a , otherwise there would be two (but one encoded).  In retrospect, I probably would have dont the initial split on , instead of !